pandas/io/pytables.py

```diff
@@ -3036,7 +3036,8 @@ def read_array(self, key: str, start: int | None = None, stop: int | None = None
             if dtype and dtype.startswith("datetime64"):
                 # reconstruct a timezone if indicated
                 tz = getattr(attrs, "tz", None)
-                ret = _set_tz(ret, tz)
+                # set time zone with parsed datetime64 unit
+                ret = _set_tz(ret, tz, unit=dtype.split("[")[1].strip("]"))
```

Comment:
I would suggest using https://numpy.org/doc/stable/reference/generated/numpy.datetime_data.html to parse the unit